packages/peregrine/src/Router/MagentoRouteHandler.js

```diff
@@ -1,74 +1,128 @@
+        const isKnown = state.componentMap.has(pathname);
 
-        if (changed && !seen) {
-            this.getRouteComponent(nextProps.location.pathname);
+        if (!isKnown) {
+            this.getRouteComponent();
         }
     }
 
-    getRouteComponent(pathname) {
-        const { apiBase, __tmp_webpack_public_path__ } = this.props;
-
-        resolveUnknownRoute({
-            route: pathname,
-            apiBase,
-            __tmp_webpack_public_path__
-        })
-            .then(({ rootChunkID, rootModuleID, matched, id }) => {
-                if (!matched) {
-                    // TODO: User-defined 404 page
-                    // when the API work is done to support it
-                    throw new Error('404');
-                }
-                return fetchRootComponent(rootChunkID, rootModuleID).then(
-                    Component => {
-                        this.setState({
-                            [pathname]: {
-                                Component,
-                                id
-                            }
-                        });
-                    }
-                );
-            })
-            .catch(err => {
-                console.log('Routing resolve failed\n', err);
+    async getRouteComponent() {
+        const { __tmp_webpack_public_path__, apiBase, location } = this.props;
+        const { pathname } = location;
+
+        try {
+            // try to resolve the route
+            // if this throws, we essentially have a 500 Internal Error
+            const resolvedRoute = await resolveUnknownRoute({
+                __tmp_webpack_public_path__,
+                apiBase,
+                route: pathname
             });
+
+            const { id, matched, rootChunkID, rootModuleID } = resolvedRoute;
+
+            // if resolution and destructuring succeed but return no match
+            // then we have a straightforward 404 Not Found
+            if (!matched) {
+                throw new Error('404');
+            }
+
+            // at this point we should have a matching RootComponent
+            // if this throws, we essentially have a 500 Internal Error
+            const RootComponent = await fetchRootComponent(
+                rootChunkID,
+                rootModuleID
+            );
+
+            // associate the matching RootComponent with this location
+            this.setRouteComponent(pathname, RootComponent, { id });
+        } catch ({ message }) {
+            const symbol = message === '404' ? NotFound : InternalError;
+
+            // we don't have a matching RootComponent, but we've checked for one
+            // so associate the appropriate error case with this location
+            this.setRouteComponent(pathname, symbol);
+        }
+    }
+
+    setRouteComponent(pathname, RootComponent, meta) {
+        this.setState(({ componentMap }) => ({
+            componentMap: new Map(componentMap).set(pathname, {
```

Comment:
How did I not know this map object existed :skull:

Reply:
It's from ES2015, so it's still kind of new. [You're one of today's lucky 10,000!](https://xkcd.com/1053/) Plain objects still have plenty of uses, but `Map` has a great API.

Here, the `.set()` method returns the `Map` itself, so it can be chained.

```es6
const foo = new Map
  .set('a', 1)
  .set('b', 2)

console.log(foo.size) // logs 2
```

Also, `Map` is [iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols), and the default iterator is the same shape that the [constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) expects, so you can quickly clone a `Map` by feeding it to the constructor as we're doing here.

```es6
const foo = new Map()
  .set('a', 1)
  .set('b', 2)

const bar = new Map(foo)

expect(bar).toEqual(foo) // pass: `foo` and `bar` contain equivalent keys and values
expect(bar).toBe(foo) // fail: `foo` and `bar` are not the same object
```